File: String to Integer (atoi).java
//Time Complexity-O(n)
//Space Complexity-O(1)
class Solution {
    public int myAtoi(String s) {
        int result=0;
        s=s.trim();
        if(s.length()==0)
        {
            return 0;
        }
        if(!Character.isDigit(s.charAt(0))&&(s.charAt(0)!='-'&&s.charAt(0)!='+'))
        {
            return 0;
        }
        char sign='+';
        if(s.charAt(0)=='-')
        {
            sign='-';
        } 
        int limit=Integer.MAX_VALUE/10;
        for(int i=0;i<s.length();i++)
        {
            if(Character.isDigit(s.charAt(i)))
            {
                if(result>=limit)
                {
                    if(sign=='+')
                    {
                        if(result>limit)
                        {
                            return Integer.MAX_VALUE;
                        }
                        else if(result==limit && (s.charAt(i)-'0'>6))
                        {
                            return Integer.MAX_VALUE;
                        }
                    }
                    else{
                        if(result>limit)
                        {
                            return Integer.MIN_VALUE;
                        }
                        else if(result==limit && (s.charAt(i)-'0'>7))
                        {
                            return Integer.MIN_VALUE;
                        }
                    }
                }
                result=result*10+(s.charAt(i)-'0');
            }
            else if(i>0)
            {
                break;
            }
        }
        if(sign=='+')
        {
            return result;
        }
        return result*-1;
    }
}

File: String to Integer (atoi).py
#time: O(n)
#space:O(1)

class Solution:
    MAX_INT = 2**31-1
    MIN_INT = -(2**31)
    def myAtoi(self, string: str) -> int:
        s = string.lstrip(' ')
        if not s:
            return 0
        hashmap={}
        for i in range(10):
            hashmap[str(i)]=i
        sign = -1 if s[0] == "-" else 1
        if sign != 1 or s[0] == "+":
            s = s[1:]
            
        res = 0
        for c in s:
            if c not in hashmap:
                return self.helper(res * sign)
            
            res *= 10
            res += hashmap[c]
            
        return self.helper(res * sign)
    
    def helper(self, x: int):
        if x > self.MAX_INT:
            return self.MAX_INT
        if x < self.MIN_INT:
            return self.MIN_INT
        return x